daemon/access_test.go

```diff
@@ -397,3 +398,65 @@ func (s *accessSuite) TestInterfaceAuthenticatedAccessPolkit(c *C) {
 	defer restore()
 	c.Check(ac.CheckAccess(s.d, req, ucred, nil), IsNil)
 }
+
+func (s *accessSuite) TestNoticeReadAccessSnap(c *C) {
+	var ac daemon.AccessChecker = daemon.NoticeReadAccess{}
+
+	s.daemon(c)
+	// interfaceOpenAccess allows access if requireInterfaceApiAccess() succeeds
+	ucred := &daemon.Ucrednet{Uid: 42, Pid: 100, Socket: dirs.SnapSocket}
+	var called int
+	restore := daemon.MockRequireInterfaceApiAccess(func(d *daemon.Daemon, u *daemon.Ucrednet, interfaceName string) *daemon.APIError {
+		called++
+		c.Check(d, Equals, s.d)
+		c.Check(u, Equals, ucred)
+		c.Check(interfaceName, Equals, "snap-refresh-observe")
+		return nil
+	})
+	defer restore()
+
+	// change-update notices can be read with snap-refresh-observe interface
+	req := http.Request{
+		URL: &url.URL{RawQuery: "types=change-update"},
+	}
+	c.Check(ac.CheckAccess(s.d, &req, ucred, nil), IsNil)
+	c.Check(called, Equals, 1)
+
+	// other notices cannot be read by a snap
+	req = http.Request{
+		URL: &url.URL{RawQuery: "types=change-update,warning"},
+	}
+	c.Check(ac.CheckAccess(s.d, &req, ucred, nil), DeepEquals, errForbidden)
+	c.Check(called, Equals, 2)
+
+	// snaps cannot read all notices
+	req = http.Request{
+		URL: &url.URL{RawQuery: ""},
+	}
+	c.Check(ac.CheckAccess(s.d, &req, ucred, nil), DeepEquals, errForbidden)
+	c.Check(called, Equals, 2)
+}
+
+func (s *accessSuite) TestNoticeReadAccessDefaultOpenAccess(c *C) {
+	var ac daemon.AccessChecker = daemon.NoticeReadAccess{}
+
+	s.daemon(c)
+	ucred := &daemon.Ucrednet{Uid: 42, Pid: 100, Socket: dirs.SnapdSocket}
+
+	// all notices are readable through snapd.socket
+	// single notice type
+	req := http.Request{
+		URL: &url.URL{RawQuery: "types=change-update"},
+	}
+	c.Check(ac.CheckAccess(s.d, &req, ucred, nil), IsNil)
+	// multiple notice types
+	req = http.Request{
+		URL: &url.URL{RawQuery: "types=change-update,warning"},
+	}
+	c.Check(ac.CheckAccess(s.d, &req, ucred, nil), IsNil)
+	// all notice types
+	req = http.Request{
+		URL: &url.URL{RawQuery: ""},
+	}
+	c.Check(ac.CheckAccess(s.d, &req, ucred, nil), IsNil)
+}
```

Comment:
Can an extra test be added which iterates through the various notice types to ensure we never get access denied when using the snapd.socket so we catch when new types get added but the `noticeReadAccess` hasn't been updated to account for them?

Reply:
good point, added [here](https://github.com/snapcore/snapd/pull/13619/files#diff-bbd4dbcc2c1898ef7e7541c852ae8d72f8d5eb3a5d37af196ddd4dde59d69aa4R449-R454)